classpath/java/io/BufferedInputStream.java

```diff
@@ -54,28 +55,7 @@ public int read(byte[] b, int offset, int length) throws IOException {
 
       count += remaining;
       position += remaining;
-      offset += remaining;
-      length -= remaining;
     }
-
-    while (length > 0) {
```

Comment:
I still think we should leave this while loop in there. Glancing at the openjdk source, they're doing something largely equivalent.  Keep in mind that the `if (in.available() <= 0)` ought to take care of your concern that we'll block on waiting for more data.